### PR TITLE
added documentation in `dynamic_programming/0_1_knapsack.cpp`

### PR DESCRIPTION
**Pull request by Pardeep009**
#### Description of Change

added documentation in `dynamic_programming/0_1_knapsack.cpp`

#### Checklist

- [x] Added description of change
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1207"></a>

### CODE REVIEW

File: dynamic_programming/0_1_knapsack.cpp [hunk 1]
```diff
@@ -1,66 +1,128 @@
...
+ * Given weights and values of n items, put these items in a knapsack of
+ * capacity W to get the maximum total value in the knapsack. In other words,
+ * given two integer arrays val[0..n-1] and wt[0..n-1] which represent values
+ * and weights associated with n items respectively. Also given an integer W
+ * which represents knapsack capacity, find out the maximum value subset of
+ * val[] such that sum of the weights of this subset is smaller than or equal to
+ * W. You cannot break an item, either pick the complete item or don’t pick it
+ * (0-1 property)
+ *
+ * ### Algorithm
```
Comment (kvedala) [inline dynamic_programming/0_1_knapsack.cpp:16]: this stub lines 16 and 17 can be removed

**Review by kvedala — APPROVED** [review]
LGTM 👍 with one minor fix

File: dynamic_programming/0_1_knapsack.cpp (hunk 1 above)
Comment (Pardeep009) [reply dynamic_programming/0_1_knapsack.cpp:16]: I have added algorithm in the comments now.

File: dynamic_programming/0_1_knapsack.cpp [hunk 2]
```diff
@@ -1,66 +1,130 @@
...
-//#include <bits/stdc++.h>
+/**
+ * @file
+ * @brief Implementation of [0-1 Knapsack Problem]
+ * (https://en.wikipedia.org/wiki/Knapsack_problem)
+ *
+ * @details
+ * Given weights and values of n items, put these items in a knapsack of
+ * capacity W to get the maximum total value in the knapsack. In other words,
+ * given two integer arrays val[0..n-1] and wt[0..n-1] which represent values
```
Comment (kvedala) [inline dynamic_programming/0_1_knapsack.cpp:9]:
```suggestion
 * given two integer arrays `val[0..n-1]` and `wt[0..n-1]` which represent values
```

File: dynamic_programming/0_1_knapsack.cpp [hunk 3]
```diff
@@ -1,66 +1,130 @@
...
+ * and weights associated with n items respectively. Also given an integer W
+ * which represents knapsack capacity, find out the maximum value subset of
+ * val[] such that sum of the weights of this subset is smaller than or equal to
```
Comment (kvedala) [inline dynamic_programming/0_1_knapsack.cpp:12]:
```suggestion
 * `val[]` such that sum of the weights of this subset is smaller than or equal to
```

File: dynamic_programming/0_1_knapsack.cpp [hunk 4]
```diff
@@ -1,66 +1,130 @@
...
+ * W. You cannot break an item, either pick the complete item or don’t pick it
```
Comment (kvedala) [inline dynamic_programming/0_1_knapsack.cpp:13]:
```suggestion
 * `W`. You cannot break an item, either pick the complete item or don’t pick it
```

File: dynamic_programming/0_1_knapsack.cpp [hunk 5]
```diff
@@ -1,66 +1,130 @@
...
+ * (0-1 property)
+ *
+ * ### Algorithm
+ * The idea is to consider all subsets of items and calculate the total weight
+ * and value of all subsets. Consider the only subsets whose total weight is
+ * smaller than W. From all such subsets, pick the maximum value subset.
```
Comment (kvedala) [inline dynamic_programming/0_1_knapsack.cpp:19]:
```suggestion
 * smaller than `W`. From all such subsets, pick the maximum value subset.
```

File: dynamic_programming/0_1_knapsack.cpp [hunk 6]
```diff
@@ -1,66 +1,130 @@
-// 0-1 Knapsack problem - Dynamic programming
-//#include <bits/stdc++.h>
+/**
+ * @file
+ * @brief Implementation of [0-1 Knapsack Problem]
+ * (https://en.wikipedia.org/wiki/Knapsack_problem)
+ *
+ * @details
+ * Given weights and values of n items, put these items in a knapsack of
+ * capacity W to get the maximum total value in the knapsack. In other words,
```
Comment (kvedala) [inline dynamic_programming/0_1_knapsack.cpp:8]:
```suggestion
 * capacity `W` to get the maximum total value in the knapsack. In other words,
```

File: dynamic_programming/0_1_knapsack.cpp [hunk 7]
```diff
@@ -1,66 +1,130 @@
...
+ * possible combinations will yield the maximum knapsack value.
+ * @tparam T size of the weight and value array
+ * @param capacity capacity of the carrying bag
+ * @param n no of items
+ * @param weight array representing weight of items
+ * @param value array representing value of items
+ * @return maximum value obtainable with given capacity.
+ */
+template <size_t T>
+int maxKnapsackValue(const int capacity, const int n,
```
Comment (kvedala) [inline dynamic_programming/0_1_knapsack.cpp:52]: This is a good update. Isn't the parameter `n` identical to the template parameter `T`? I think the input argument `n` can be eliminated using something as this:
```cpp
template <size_t n>
 int maxKnapsackValue(const int capacity,
                      const std::array<int, n> &weight,
                      const std::array<int, n> &value) {
```

**Review by kvedala — CHANGES_REQUESTED** [review]

File: dynamic_programming/0_1_knapsack.cpp (hunk 7 above)
Comment (Pardeep009) [reply dynamic_programming/0_1_knapsack.cpp:52]: Right, I will update it.

**Comment by Pardeep009** [comment]
@kvedala can you review my code?

**Review by kvedala — APPROVED** [review]
well done 👍

File: dynamic_programming/0_1_knapsack.cpp [hunk 8]
```diff
@@ -1,66 +1,129 @@
...
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @namespace Knapsack
+ * @brief Implementation of 0-1 Knapsack problem
+ */
+namespace Knapsack {
```
Comment (Panquesito7) [inline dynamic_programming/0_1_knapsack.cpp:39]: I think this shall be non-capitalized letter.
```suggestion
namespace knapsack {
```

File: dynamic_programming/0_1_knapsack.cpp [hunk 9]
```diff
@@ -1,66 +1,129 @@
...
         }
     }
-    //	Print(res, n, capacity, capacity);
-    //	cout<<"\n";
-    return res[n][capacity];
+
+    // returning maximum value
+    return maxValue[items][capacity];
+}
+}  // namespace Knapsack
```
Comment (Panquesito7) [inline dynamic_programming/0_1_knapsack.cpp:88]:
```suggestion
}  // namespace knapsack
```

File: dynamic_programming/0_1_knapsack.cpp [hunk 10]
```diff
@@ -1,66 +1,129 @@
...
+ * @brief Function to test above algorithm
+ * @returns void
+ */
+static void test() {
+    // Test 1
+    const int n1 = 3;                             // number of items
+    std::array<int, n1> weight1 = {10, 20, 30};   // weight of each item
+    std::array<int, n1> value1 = {60, 100, 120};  // value of each item
+    const int capacity1 = 50;                     // capacity of carrying bag
+    const int max_value1 = dynamic_programming::Knapsack::maxKnapsackValue(
```
Comment (Panquesito7) [inline dynamic_programming/0_1_knapsack.cpp:101]:
```suggestion
    const int max_value1 = dynamic_programming::knapsack::maxKnapsackValue(
```

File: dynamic_programming/0_1_knapsack.cpp [hunk 11]
```diff
@@ -1,66 +1,129 @@
...
+    assert(max_value1 == expected_max_value1);
+    std::cout << "Maximum Knapsack value with " << n1 << " items is "
+              << max_value1 << std::endl;
+
+    // Test 2
+    const int n2 = 4;                               // number of items
+    std::array<int, n2> weight2 = {24, 10, 10, 7};  // weight of each item
+    std::array<int, n2> value2 = {24, 18, 18, 10};  // value of each item
+    const int capacity2 = 25;                       // capacity of carrying bag
+    const int max_value2 = dynamic_programming::Knapsack::maxKnapsackValue(
```
Comment (Panquesito7) [inline dynamic_programming/0_1_knapsack.cpp:113]:
```suggestion
    const int max_value2 = dynamic_programming::knapsack::maxKnapsackValue(
```

**Review by Panquesito7 — APPROVED** [review]
Amazing work, I like how you've improved the code so well.
Code and documentation looks good and very refined; LGTM. 😄 👍

**Review by ayaankhan98 — APPROVED** [review]
good work! thanks for your contribution

**Review by ayaankhan98 — APPROVED** [review]
Nice Improvement! LGTM!

**Review by Panquesito7 — APPROVED** [review]
👍

**Pull request merged**